inc/saitam.h

```diff
@@ -34,866 +34,2191 @@
 
 /**
- * @brief TAM Snapshot Transporter Attributes
+ * @brief TAM collector attributes
  */
-typedef enum _sai_tam_transporter_attr_t
+typedef enum _sai_tam_collector_attr_t
 {
+
     /**
      * @brief Start of Attributes
      */
-    SAI_TAM_TRANSPORTER_ATTR_START,
+    SAI_TAM_COLLECTOR_ATTR_START,
 
     /**
-     * @brief Transporter Type
+     * @brief Source IP address
      *
-     * If this attribute value is unspecified the local transporter is used.
+     * Note: Applicable only when SAI_TAM_TRANSPORT_ATTR_TRANSPORT_TYPE != SAI_TAM_TRANSPORT_TYPE_NONE
      *
-     * @type sai_tam_transporter_type_t
-     * @flags CREATE_AND_SET
-     * @default SAI_TAM_TRANSPORTER_TYPE_LOCAL
+     * @type sai_ip_address_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
      */
-    SAI_TAM_TRANSPORTER_ATTR_TYPE = SAI_TAM_TRANSPORTER_ATTR_START,
+    SAI_TAM_COLLECTOR_ATTR_SRC_IP = SAI_TAM_COLLECTOR_ATTR_START,
 
     /**
-     * @brief Maximum size beyond which it will be truncated.
+     * @brief Destination IP addresses
      *
-     * If this attribute value is zero or unspecified, snapshots are not
-     * truncated while transporting. (DEFAULT)
+     * Note: Applicable only when SAI_TAM_TRANSPORT_ATTR_TRANSPORT_TYPE != SAI_TAM_TRANSPORT_TYPE_NONE
      *
-     * @type sai_uint32_t
-     * @flags CREATE_AND_SET
-     * @default 0
+     * @type sai_ip_address_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
      */
-    SAI_TAM_TRANSPORTER_ATTR_MAX_SNAPSHOT_SIZE,
+    SAI_TAM_COLLECTOR_ATTR_DST_IP,
 
     /**
-     * @brief Mirroring session object defining the remote transport capabilities.
+     * @brief Destination local CPU
      *
-     * If this attribute is unspecified, Local CPU Transport is used (DEFAULT).
+     * Note: Applicable only when SAI_TAM_TRANSPORT_ATTR_TRANSPORT_TYPE == SAI_TAM_TRANSPORT_TYPE_NONE
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default true
+     */
+    SAI_TAM_COLLECTOR_ATTR_LOCALHOST,
+
+    /**
+     * @brief Virtual router ID
      *
      * @type sai_object_id_t
      * @flags CREATE_AND_SET
-     * @objects SAI_OBJECT_TYPE_MIRROR_SESSION
+     * @objects SAI_OBJECT_TYPE_VIRTUAL_ROUTER
      * @allownull true
      * @default SAI_NULL_OBJECT_ID
      */
-    SAI_TAM_TRANSPORTER_ATTR_MONITOR_ID,
+    SAI_TAM_COLLECTOR_ATTR_VIRTUAL_ROUTER_ID,
+
+    /**
+     * @brief Telemetry report truncate size
+     *
+     * @type sai_uint16_t
+     * @flags CREATE_AND_SET
+     * @isvlan false
+     * @default 0
+     */
+    SAI_TAM_COLLECTOR_ATTR_TRUNCATE_SIZE,
+
+    /**
+     * @brief Transport attributes object
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TAM_TRANSPORT
+     */
+    SAI_TAM_COLLECTOR_ATTR_TRANSPORT,
+
+    /**
+     * @brief DSCP value
+     *
+     * @type sai_uint8_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_TAM_COLLECTOR_ATTR_DSCP_VALUE,
```

Comment:
Quoting from Jai's reply:

> [Jai] Assigning each data set a DSCP value is not typical deployment. Typically collector subsystem is scaled differently (look at openconfig telemetry semantics). Each session (aka virtual collector or session can be a collector) is assigned a DSCP value for all the telemetry traffic carried over that session. Multiple sessions (or collectors) are provisioned to carry different class of data. Typically Low/Medium/High Frequency class collectors/sessions are configured. High frequency data like queue stats, medium frequency data like flow stats, low frequency like ports stats. Events are usually send on a stateful channel and need not have a different DSCP value. If events are send on a UDP channel then even assigning them high DSCP value doesn’t guarantee the delivery,

Reply:
This is not about guaranteeing delivery, it is about prioritization between different types of telemetry traffic. As telemetry use cases and mechanisms proliferate, use of telemetry is increasing. Along with that comes an increase in the amount of telemetry traffic, which may lead to congestion.

We are defining a new mechanism here. Shouldn't we be anticipating potential problems and improving functionality, rather than looking to what has been done in the past?

Forcing separate TAM objects, with separate instances of the entire TAM hierarchy including binding and collectors, just to get different DSCP values in different TAM reports is more than a little unwieldy.